File: app/src/main/java/co/tpcreative/suppersafe/common/adapter/DividerItemDecoration.java
package co.tpcreative.suppersafe.common.adapter;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

public class DividerItemDecoration extends RecyclerView.ItemDecoration {

    private static final int[] ATTRS = new int[]{
            android.R.attr.listDivider
    };

    public static final String TAG = DividerItemDecoration.class.getSimpleName();

    public static final int HORIZONTAL_LIST = LinearLayoutManager.HORIZONTAL;

    public static final int VERTICAL_LIST = LinearLayoutManager.VERTICAL;

    private Drawable mDivider;

    private int mOrientation;

    private int mMarginTop = 0;
    private int mMarginBottom = 0;
    private int mMarginLeft = 0;
    private int mMarginRight = 0;

    public DividerItemDecoration(Context context, int orientation) {
        final TypedArray a = context.obtainStyledAttributes(ATTRS);
        if (mDivider == null){
            mDivider = a.getDrawable(0);
        }
        a.recycle();
        setOrientation(orientation);
    }

    public void setDrawable(Drawable mResource){
        this.mDivider = mResource;
    }
    public Drawable getDivider(){
        return mDivider;
    }

    public void setOrientation(int orientation) {
        if (orientation != HORIZONTAL_LIST && orientation != VERTICAL_LIST) {
            throw new IllegalArgumentException("invalid orientation");
        }
        mOrientation = orientation;
    }

    @Override
    public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
        if (mOrientation == VERTICAL_LIST) {
            drawVertical(c, parent);
        } else {
            drawHorizontal(c, parent);
        }
    }

    public void drawVertical(Canvas c, RecyclerView parent) {
        final int left = parent.getPaddingLeft();
        final int right = parent.getWidth() - parent.getPaddingRight();

        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child
                    .getLayoutParams();
            final int top = child.getBottom() + params.bottomMargin;
            final int bottom = top + mDivider.getIntrinsicHeight();

            mDivider.setBounds(left,top,right,bottom);
            mDivider.draw(c);
        }
    }

    public void drawHorizontal(Canvas c, RecyclerView parent) {
        final int top = parent.getPaddingTop();
        final int bottom = parent.getHeight() - parent.getPaddingBottom();

        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child
                    .getLayoutParams();
            final int left = child.getRight() + params.rightMargin;
            final int right = left + mDivider.getIntrinsicHeight();
            mDivider.setBounds(left,top,right,bottom);
            mDivider.draw(c);
        }
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        if (mOrientation == VERTICAL_LIST) {
            outRect.set(0, 0, 0, mDivider.getIntrinsicHeight());
        } else {
            outRect.set(0, 0, mDivider.getIntrinsicWidth(), 0);
        }
    }

    public void setMarginTop(int marginTop){
        this.mMarginTop = marginTop ;
    }
    public int getMarginTop(){
        return mMarginTop;
    }

    public void setMarginBottom(int marginBottom){
        this.mMarginBottom = marginBottom ;
    }
    public int getMarginBottom(){
        return mMarginBottom;
    }

    public void setMarginLeft(int mMarginLeft){
        this.mMarginLeft = mMarginLeft ;
    }
    public int getMarginLeft(){
        return mMarginLeft;
    }

    public void setMarginRight(int marginRight){
        this.mMarginRight = marginRight ;
    }

    public int getMarginRight(){
        return mMarginRight;
    }





}

File: app/src/main/java/co/tpcreative/supersafe/common/adapter/DividerItemDecoration.kt
package co.tpcreative.supersafe.common.adapter
import android.R
import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import co.tpcreative.supersafe.common.adapter.DividerItemDecoration

class DividerItemDecoration(context: Context?, orientation: Int) : RecyclerView.ItemDecoration() {
    private var mDivider: Drawable? = null
    private var mOrientation = 0
    private var mMarginTop = 0
    private var mMarginBottom = 0
    private var mMarginLeft = 0
    private var mMarginRight = 0
    fun setDrawable(mResource: Drawable?) {
        mDivider = mResource
    }

    fun getDivider(): Drawable? {
        return mDivider
    }

    fun setOrientation(orientation: Int) {
        require(!(orientation != DividerItemDecoration.Companion.HORIZONTAL_LIST && orientation != DividerItemDecoration.Companion.VERTICAL_LIST)) { "invalid orientation" }
        mOrientation = orientation
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (mOrientation == DividerItemDecoration.Companion.VERTICAL_LIST) {
            drawVertical(c, parent)
        } else {
            drawHorizontal(c, parent)
        }
    }

    fun drawVertical(c: Canvas?, parent: RecyclerView) {
        val left: Int = parent.getPaddingLeft()
        val right: Int = parent.getWidth() - parent.getPaddingRight()
        val childCount: Int = parent.getChildCount()
        for (i in 0 until childCount) {
            val child: View = parent.getChildAt(i)
            val params: RecyclerView.LayoutParams = child
                    .layoutParams as RecyclerView.LayoutParams
            val top: Int = child.bottom + params.bottomMargin
            val bottom = top + mDivider?.getIntrinsicHeight()!!
            mDivider?.setBounds(left, top, right, bottom)
            if (c != null) {
                mDivider?.draw(c)
            }
        }
    }

    fun drawHorizontal(c: Canvas?, parent: RecyclerView) {
        val top: Int = parent.getPaddingTop()
        val bottom: Int = parent.getHeight() - parent.getPaddingBottom()
        val childCount: Int = parent.getChildCount()
        for (i in 0 until childCount) {
            val child: View = parent.getChildAt(i)
            val params: RecyclerView.LayoutParams = child
                    .layoutParams as RecyclerView.LayoutParams
            val left: Int = child.right + params.rightMargin
            val right = left + mDivider!!.getIntrinsicHeight()
            mDivider?.setBounds(left, top, right, bottom)
            if (c != null) {
                mDivider?.draw(c)
            }
        }
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        if (mOrientation == DividerItemDecoration.Companion.VERTICAL_LIST) {
            outRect.set(0, 0, 0, mDivider!!.getIntrinsicHeight())
        } else {
            outRect.set(0, 0, mDivider!!.getIntrinsicWidth(), 0)
        }
    }

    fun setMarginTop(marginTop: Int) {
        mMarginTop = marginTop
    }

    fun getMarginTop(): Int {
        return mMarginTop
    }

    fun setMarginBottom(marginBottom: Int) {
        mMarginBottom = marginBottom
    }

    fun getMarginBottom(): Int {
        return mMarginBottom
    }

    fun setMarginLeft(mMarginLeft: Int) {
        this.mMarginLeft = mMarginLeft
    }

    fun getMarginLeft(): Int {
        return mMarginLeft
    }

    fun setMarginRight(marginRight: Int) {
        mMarginRight = marginRight
    }

    fun getMarginRight(): Int {
        return mMarginRight
    }

    companion object {
        private val ATTRS: IntArray? = intArrayOf(
                R.attr.listDivider
        )
        val TAG = DividerItemDecoration::class.java.simpleName
        val HORIZONTAL_LIST: Int = androidx.recyclerview.widget.LinearLayoutManager.HORIZONTAL
        val VERTICAL_LIST: Int = androidx.recyclerview.widget.LinearLayoutManager.VERTICAL
    }

    init {
        val a: TypedArray? = DividerItemDecoration.ATTRS?.let { context!!.obtainStyledAttributes(it) }
        if (mDivider == null) {
            mDivider = a?.getDrawable(0)
        }
        a?.recycle()
        setOrientation(orientation)
    }
}